libyul/backends/evm/SSACFGTopologicalSort.cpp

```diff
@@ -0,0 +1,89 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libyul/backends/evm/SSACFGTopologicalSort.h>
+
+using namespace solidity::yul;
+
+ForwardSSACFGTopologicalSort::ForwardSSACFGTopologicalSort(SSACFG const& _cfg):
+	m_cfg(_cfg),
+	m_explored(m_cfg.numBlocks(), false), m_blockWisePreOrder(m_cfg.numBlocks(), 0),
+	m_blockWiseMaxSubtreePreOrder(m_cfg.numBlocks(), 0)
+{
+	yulAssert(m_cfg.entry.value == 0);
+	m_preOrder.reserve(m_cfg.numBlocks());
+	m_postOrder.reserve(m_cfg.numBlocks());
+	for (size_t id = 0; id < m_cfg.numBlocks(); ++id)
+	{
+		if (!m_explored[id])
+			dfs(id);
+	}
+
+	for (auto const& [v1, v2]: m_potentialBackEdges)
+		if (ancestor(v2, v1))
+			m_backEdgeTargets.insert(v2);
+}
+
+void ForwardSSACFGTopologicalSort::dfs(size_t const _vertex) {
+	yulAssert(!m_explored[_vertex]);
+	m_explored[_vertex] = true;
+	m_blockWisePreOrder[_vertex] = m_preOrder.size();
+	m_blockWiseMaxSubtreePreOrder[_vertex] = m_blockWisePreOrder[_vertex];
+	m_preOrder.push_back(_vertex);
+
+	m_cfg.block(SSACFG::BlockId{_vertex}).forEachExit([&](SSACFG::BlockId const& _exitBlock){
+		if (!m_explored[_exitBlock.value])
+		{
+			dfs(_exitBlock.value);
+			m_blockWiseMaxSubtreePreOrder[_vertex] = std::max(m_blockWiseMaxSubtreePreOrder[_vertex], m_blockWiseMaxSubtreePreOrder[_exitBlock.value]);
+		}
+		else
+			m_potentialBackEdges.emplace_back(_vertex, _exitBlock.value);
+	});
+
+	m_postOrder.push_back(_vertex);
+}
+
+std::vector<size_t> const& ForwardSSACFGTopologicalSort::preOrder() const { return m_preOrder; }
+std::vector<size_t> const& ForwardSSACFGTopologicalSort::postOrder() const { return m_postOrder; }
+std::set<size_t> const& ForwardSSACFGTopologicalSort::backEdgeTargets() const { return m_backEdgeTargets; }
+SSACFG const& ForwardSSACFGTopologicalSort::cfg() const { return m_cfg; }
```

Comment:
I didn't give it much thought at the time. So now, after some thinking, here are my considerations: 

- It would lead to more inlining opportunities for the compiler (if it makes a difference, stuff like is also picked up by LTO - do we have it enabled?). So if this was used in a tight loop then it might very well make a performance difference (which has to be profiled, doubt it would be a big one).
- On the other hand, doing it like this can bring down compile times and if something has to be changed in the implementation, incremental builds are faster. Although that seems very unlikely in this case unless the name of the member changes, in which case the header changes as well.
- Also it fosters uniformity and a better declaration/definition separation.

Bottom line: It very likely doesn't matter in this case from a runtime performance point of view, would have to be profiled in any case. It probably increases compile time performance very slightly (but also negligible b/c it's not a header that would be included in a lot of places or very high up the include hierarchy). Meh :)

Reply:
I've put them in the header now :laughing:

Reply:
I was more curious than anything here :)
I thought about the first two points you mentioned and was wondering if there were any other considerations.
Thanks for the explanation @clonker !